### PR TITLE
Projects: set up table variants for (un)publishing via buttons vs. dropdown

### PR DESCRIPTION
**Pull request by Erin007**
There is some concern that publishing from the quick actions dropdown via the chevron icon will be less discoverable than publishing via the buttons directly in the table (currently in the Angular version of the table).  We want to ensure that students continue to publish projects with the same or greater frequency with the new React table. Therefore, we're going to run an A/B experiment to compare (un)publishing from the quick actions dropdown and from a button in the published column.

Variant A  (`publishMethod = 'chevron'`) will be the same as the current behavior in the React version of the table. (Un)publishing will take place via a quick action in the dropdown menu accessible via the chevron icon. From Storybook: 
screen shot 2018-08-13 at 1 10 24 pm

Variant B  (`publishMethod = 'button'`) will be similar to the current behavior in the Angular version of the table. (Un)publishing will take place via (un)publish buttons displayed in the Published column of the table. From Storybook: 
screen shot 2018-08-13 at 1 59 58 pm

To accomplish this change, I extracted `PersonalProjectsPublishedCell` into a separate component, passed a new 'publishMethod` prop to the `PersonalProjectsTable` and included Stories to cover the variants. 

The React table in use (still behind a flag) currently has `publishMethod = 'chevron'`.  The next step will be to show the different variants to users based on whether their user_id is even or odd, then to log publishing rates for each group.

### CODE REVIEW

**Comment by Erin007** [comment]
cc: @poorvasingal Do these versions of the table match what you were imagining?

**Comment by davidsbailey** [comment]
Apologies for being the reviewer that creates more work, but if you really want more projects to be published, then how about:
screen shot 2018-08-13 at 2 22 11 pm

File: apps/src/templates/projects/PersonalProjectsPublishedCell.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import {showPublishDialog} from './publishDialog/publishDialogRedux';
+import {unpublishProject} from './projectsRedux';
+
+class PersonalProjectsPublishedCell extends Component {
+  static propTypes = {
+    projectId: PropTypes.string.isRequired,
+    projectType: PropTypes.string.isRequired,
+    isPublished: PropTypes.bool.isRequired,
+    isPublishable: PropTypes.bool.isRequired,
+    publishMethod: PropTypes.oneOf(['button', 'chevron']).isRequired,
```
Comment (davidsbailey) [inline apps/src/templates/projects/PersonalProjectsPublishedCell.jsx]: these would be best as constants exported by this file:
```js
export const publishMethods = {
  PUBLISH: 'publish',
  CHEVRON: 'chevron',
}
```
and then referenced as `publishMethods.CHEVRON`, here and in PersonalProjectsTable.jsx .

File: apps/src/templates/projects/PersonalProjectsTable.jsx [hunk 2]
```diff
@@ -114,14 +114,13 @@ const dateFormatter = function (time) {
...
-  return publishedAt ? (<FontAwesome icon="check"/>) : '';
-};
-
 class PersonalProjectsTable extends React.Component {
   static propTypes = {
     personalProjectsList: PropTypes.arrayOf(personalProjectDataPropType).isRequired,
-    canShare: PropTypes.bool.isRequired
+    canShare: PropTypes.bool.isRequired,
+    // We're going to run an A/B experiment to compare (un)publishing from the // quick actions dropdown and from a button in the published column.
+    // TODO (Erin B.) delete this prop and the less effective variant when we // determine the experiment outcome.
```
Comment (davidsbailey) [inline apps/src/templates/projects/PersonalProjectsTable.jsx]: nit: add newlines before trailing `//`s

**Comment by Erin007** [comment]
Interesting idea, @davidsbailey! I like that it promotes publishing over unpublishing, but have tiny concern that it might be confusing to publish and unpublish in different places. Not sure how valid that concern is though. I'll check with Poorva.

File: apps/src/templates/projects/PersonalProjectsTable.jsx [hunk 3]
```diff
@@ -131,14 +130,33 @@ class PersonalProjectsTable extends React.Component {
     }
   };
 
+  publishedAtFormatter = (publishedAt, {rowData}) => {
+    const {canShare, publishMethod} = this.props;
+    const isPublishable =
+      AlwaysPublishableProjectTypes.includes(rowData.type) ||
+      (ConditionallyPublishableProjectTypes.includes(rowData.type) && canShare);
```
Comment (davidsbailey) [inline apps/src/templates/projects/PersonalProjectsTable.jsx:140]: this code is duplicated in `actionsFormatter` below. Please extract a helper

File: apps/src/templates/projects/PersonalProjectsPublishedCell.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+export default connect(state => ({}), dispatch => ({
+  showPublishDialog(projectId, projectType) {
+    dispatch(showPublishDialog(projectId, projectType));
+  },
+  unpublishProject(projectId) {
+    dispatch(unpublishProject(projectId));
+  },
+}))(PersonalProjectsPublishedCell);
```
Comment (davidsbailey) [inline apps/src/templates/projects/PersonalProjectsPublishedCell.jsx:66]: nice extraction!

**Comment by poorvasingal** [comment]
What you have in the original screenshots matches what I was expecting. I agree with your thoughts on moving where the publish option is might be confusing. Also part of the goal with the A/B experiment is to see how much more discoverable a button outside the chevron feels - something I'd like to extrapolate for other tables too (like the sections table).

**Comment by davidsbailey** [comment]
Fair point about consistency, however I will point out that it's a little confusing to have a Published column filled with "Publish" and "Unpublish", which are showing the opposite of the current state of each project, whereas a mix of check marks and Publish buttons makes it a bit more clear.

I'm fine with whatever you decide to launch for the experiment. Perhaps if the experiment shows a difference in publish button clicks, the screenshot I pasted can be reconsidered as an option for what we ship.

**Comment by Erin007** [comment]
Thanks for the thoughts - my inclination is to wait and see the publish and unpublish rates in each scenario with both unpublish and publish as buttons and unpublish and publish in the dropdown.  If we find that people are publishing more from the button than from dropdown, then we can follow up and compare the button scenario to Dave's hybrid scenario and see if it reduces unpublishing/confuses users. Hooray for iteration!

**Pull request merged**